Control plane EE

jbradberry opened the pull request:
##### SUMMARY

Per discussion, we are now going to have privileged execution environments for use on the control plane, and these will not be allowed to be edited or deleted.

related https://github.com/ansible/tower/issues/5016

##### TODO

- [x] reinstate the restrictive RBAC for `managed_by_tower=True` EEs
- [x] convert any EEs automatically installed prior into `managed_by_tower=False` EEs
- [x] change the resolver so that ordinary jobs do not get a managed EE, and project updates only get a managed EE
- [x] allow sysadmin users to edit the `pull` field for managed EEs 
- [x] automatically disassociate EEs that get deleted from the `DEFAULT_EXECUTION_ENVIRONMENT` setting #10363

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - API

jbradberry committed:
Remove the managed flag from all existing EEs

This flag henceforth is going to be used only for the "control plane"
execution environments, which sysadmins will not be allowed to alter.

jbradberry committed:
Update the EE resolver logic

so that the control plane managed EE is kept separate.

AlanCoding commented:
Could we merge this with #10377 from @kdelee and go from there?

jbradberry commented:
I don't see why this needs to be merged with another PR before landing.  These changes are pretty distinct.

AlanCoding commented on awx/main/utils/execution_environments.py:
```diff
@@ -6,10 +6,14 @@
 
 
+def get_control_plane_execution_environment():
+    return ExecutionEnvironment.objects.filter(organization=None, managed_by_tower=True).first()
```
Perhaps it's already established as a TODO, but since the pre-load EEs are concretely going to be a list with >1 elements, we need a selection criteria other than `.first()`

jbradberry replied:
Based on the ordering criteria I imposed on the EE model way back when, `.first()` translates to 'most recent', which I think is the correct priority we should give to installed EEs.

AlanCoding replied:
intended use (near-term) is that someone passes a list of 4 EEs to create in the installer

the `awx-manage` command loops over those and creates them

what you're saying would indicate that jobs will use the _final_ EE in the list as the default. This is high-value information, so much so that I will @Spredzy since he may be implementing the installer side of this.

jbradberry replied:
Yes, though the idea originally was that newer EEs installed when updating would be preferentially used over prior ones.

AlanCoding commented on awx/main/models/projects.py:
```diff
@@ -185,11 +185,11 @@ def clean_credential(self):
         are not flexible enough to allow customizing the image they use.
         """
-        return get_default_execution_environment()
+        return get_control_plane_execution_environment()
```
This makes it so project updates run with the "control plane EE".

Should ec2 inventory updates also run with the "control plane EE"? Why or why not?

jbradberry replied:
I don't have a strong opinion about it, but presumably there was a reason we hadn't had that in place previously.  @shanemcd ?

shanemcd replied:
- Inventory updates don't run on the control plane. 
- Users are able to choose which EE they want to use at the inventory source level.

jbradberry committed:
Allow sysadmins to be able to change the pull field for managed EEs

jbradberry committed:
Add a signal handler to remove the default EE if it gets deleted

jbradberry committed:
Make sure that managed EEs can't be deleted

jbradberry committed:
Deal with the possibility of get_default_pod_spec not finding an EE

shanemcd commented on awx/api/views/__init__.py:
```diff
@@ -695,17 +695,30 @@ class TeamAccessList(ResourceAccessList):
+                left = getattr(instance, field, None)
+                right = request.data.get(field, None)
+                if left is not None and right is not None and left != right:
+                    raise PermissionDenied(_("Only the 'pull' field can be edited for managed execution environments."))
```
❤️

shanemcd commented on awx/main/signals.py:
```diff
@@ -623,6 +624,12 @@ def deny_orphaned_approvals(sender, instance, **kwargs):
         approval.deny()
 
 
+@receiver(pre_delete, sender=ExecutionEnvironment)
```
🙌

jbradberry replied:
Turns out that this isn't sufficient :(

jbradberry replied:
fixed.

jbradberry commented on awx/api/views/__init__.py:
```diff
@@ -695,17 +695,30 @@ class TeamAccessList(ResourceAccessList):
+            for field in fields_to_check:
+                left = getattr(instance, field, None)
+                right = request.data.get(field, None)
+                if left is not None and right is not None and left != right:
```
This might be too much, I may need to do just `left != right` in order to avoid things like setting the Org.

jbradberry committed:
Fix a problem with using PrimaryKeyRelatedField in our settings registry

DRF, when using this field, short-circuits the call to
.to_representation() when the value is None, since clearly you aren't
going to be able to get the .pk attribute off of it in that case.  We
were previously unconditionally calling .to_representation() which
throws an error when we try to clear the value of
DEFAULT_EXECUTION_ENVIRONMENT.

shanemcd committed:
Only install ansible-core for collection tests

Faster!

shanemcd committed:
Remove usage of AWXReceptorJob in metadata.py

shanemcd committed:
Fix tests

jbradberry committed:
Remove unused import

nixocio commented:
I got this code locally, and we still allow the managed by tower EE to be deleted by system admin as per summary fields.

![image](https://user-images.githubusercontent.com/9053044/121417765-3f29fd00-c938-11eb-82ca-3940c3dacf39.png)

jbradberry committed:
Make sure that the delete capability isn't hardcoded to be the same as edit
